Rewrite of the configuration reference guide (#16496)
quarkusio/quarkus: docs/src/main/asciidoc/config-reference.adoc

:sectnumlevels: 4
:toc:

> IMPORTANT: The content of this guide and been revised and split into additional topics. Please check the <<additional-information,Additional Information>> section.

Reviewer: ```suggestion
IMPORTANT: The content of this guide has been revised and split into additional topics. Please check the <<additional-information,Additional Information>> section.
```

TIP: If you're looking for information how to make a Quarkus extension configurable then see the
<<writing-extensions.adoc#configuration,Writing Your Own Extension>> guide.

[[configuration-sources]]
== Config Sources

By default, Quarkus reads configuration properties from multiple sources (by descending ordinal):

1. (400) <<system-properties,System properties>>
2. (300) <<environment-variables,Environment variables>>
3. (295) <<env-file,.env>> file in the current working directory
4. (250) <<application-properties-file,Quarkus Application configuration file>> `application.properties` in classpath and/or
`$PWD/config/application.properties`
5. (100) <<microprofile-config-properties-file,MicroProfile Config configuration file>>
`META-INF/microprofile-config.properties` in classpath

The final configuration is the aggregation of the properties defined by all these sources. A configuration property
lookup starts by the highest ordinal configuration source available and works it way down to other sources until a
match is found. This means that any configuration property may override a value just by setting a different value in a

> higher ordinal config source. For example, a property configured using an environment property overrides the value p
> rovided using the `application.properties file.

Reviewer: ```suggestion
higher ordinal config source. For example, a property configured using an environment property overrides the value
provided using the `application.properties` file.
```

image::config-sources.png[]

[[system-properties]]
=== System properties

* for a runner jar: `java -Dquarkus.datasource.password=youshallnotpass -jar target/quarkus-app/quarkus-run.jar`
* for a native executable: `./target/myapp-runner -Dquarkus.datasource.password=youshallnotpass`

[[environment-variables]]
=== Environment variables

* for a runner jar: `export QUARKUS_DATASOURCE_PASSWORD=youshallnotpass ; java -jar target/quarkus-app/quarkus-run.jar`
* for a native executable: `export QUARKUS_DATASOURCE_PASSWORD=youshallnotpass ; ./target/myapp-runner`

NOTE: Environment variables names follow the conversion rules specified by
link:https://github.com/eclipse/microprofile-config/blob/master/spec/src/main/asciidoc/configsources.asciidoc#default-configsources[MicroProfile Config].

[[env-file]]
=== `.env` file in the current working directory

..env
[source,properties]
----
QUARKUS_DATASOURCE_PASSWORD=youshallnotpass <1>
----

> <1> The name `QUARKUS_DATASOURCE_PASSWORD` the same conversion rules used for <<environment_variables>>.

Reviewer: ```suggestion
<1> The name `QUARKUS_DATASOURCE_PASSWORD` the same conversion rules used for <<environment-variables>>.
```
